[ONNX] Support constant tensors in FakeMode exporting

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107836

Fixes https://github.com/pytorch/pytorch/issues/107475

- Constant tensors was wrongly recognized as weights and buffers, and then was detached from its default value during `to_model_proto`. This PR fixes the bug and pick up Bloom CI test back successfully. NOTE: non-persistent buffer and weights has different situation and is not fixed by this PR.
- Reduce transformers model size by modifying their config parameters to speed up CI tests. (Unrelated to this PR title)

Corresponding change with https://github.com/microsoft/onnxscript/pull/1023

titaiwangms committed:
[ONNX] FiSupport constant tensors in FakeMode exporting

[ghstack-poisoned]

titaiwangms commented on test/onnx/test_fx_to_onnx.py:
```diff
@@ -386,21 +386,14 @@ def forward(self, x):
                 )
 
-            # Scenario 2: Fake model and real input WITHOUT fake_context
-            with self.assertRaises(torch.onnx.OnnxExporterError):
```
This scenario relys the error from `to_model_proto` crashing with fake tensor initializer which does not happen anymore after the filtering logic.

BowenBao replied:
How is it affected by this PR. weight and bias of `self.linear` should be parameter and hence still fake tensor right?

titaiwangms replied:
Because previous to this PR, in this case, we don't actiavte fake_mode during the export, which triggers `to_model_proto` takes `include_initializers=True`, so we send FakeTensor into `_export_onnx`. 

But now, with the fix, this will never be the case. If we really want to error this case out, we could add Assertion into `_assert_fake_mode` for consistency, but for now, this case is actually supported.

BowenBao replied:
Discussed offline, to be brought back.

justinchuby commented on torch/onnx/_internal/exporter.py:
```diff
@@ -915,10 +915,17 @@ def export(self) -> ExportOutput:
+            for initializer_name, initializer in onnxscript_graph.initializers.items():
+                if not isinstance(initializer, torch._subclasses.FakeTensor):
+                    initializers_with_real_tensors[initializer_name] = initializer
+            onnxscript_graph.initializers = initializers_with_real_tensors
```
I would just call add for each initializer

titaiwangms replied:
They are already in it. We just need another iteration to filter out the ones with fake tensor before sending them to `to_model_proto`.

titaiwangms replied:
Currently, they are all filtered out by `include_initializers`, if fake mode is activated. That creates the bug because constant tensors do need their value in initializers.

titaiwangms replied:
If you don't like the setter, I guess the other way around would be we change `include_initializers` to something like `filtered_initializers`, and passed in with `to_model_proto` API.

titaiwangms replied:
Or I guess we could add another parameter: fake_tensor: bool in `add_initializer`.

titaiwangms committed:
Update on "[ONNX] Support constant tensors in FakeMode exporting"


Fixes https://github.com/pytorch/pytorch/issues/107475

- Constant tensors was wrongly recognized as weights and buffers, and then was detached from its default value during `to_model_proto`. This PR fixes the bug and pick up Bloom CI test back successfully. NOTE: non-persistent buffer and weights has different situation and is not fixed by this PR.
- Reduce transformers model size by modifying their config parameters to speed up CI tests.

Corresponding change with https://github.com/microsoft/onnxscript/pull/1023

[ghstack-poisoned]

titaiwangms commented on test/onnx/test_fx_to_onnx.py:
```diff
@@ -411,7 +404,9 @@ def forward(self, x):
     def test_fake_tensor_mode_huggingface_gpt2(self):
-        config = transformers.GPT2Config()
+        config = transformers.GPT2Config(
+            vocab_size=8096, n_positions=256, n_embd=256, n_layer=2, n_head=2
```
Not really. I just found they are taking a lot of time on running, so I trimmed a little bit.

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -960,6 +953,12 @@ def _assert_fake_tensor_mode(self):
+        if (
+            has_any_fake_tensor or has_any_fake_param_or_buffer
+        ) and not self.options.fake_context:
+            raise RuntimeError(
```
Related to the comment above, I suppose this should still raise for that case?

titaiwangms replied:
Yes, I could move it back. I just found it's hard to pdb it when it's separate so far lol.

titaiwangms committed:
Update on "[ONNX] Support constant tensors in FakeMode exporting"


Fixes https://github.com/pytorch/pytorch/issues/107475

- Constant tensors was wrongly recognized as weights and buffers, and then was detached from its default value during `to_model_proto`. This PR fixes the bug and pick up Bloom CI test back successfully. NOTE: non-persistent buffer and weights has different situation and is not fixed by this PR.
- Reduce transformers model size by modifying their config parameters to speed up CI tests. (Unrelated to this PR title)

Corresponding change with https://github.com/microsoft/onnxscript/pull/1023

[ghstack-poisoned]

BowenBao commented on torch/onnx/_internal/fx/fx_onnx_interpreter.py:
```diff
@@ -806,14 +806,19 @@ def get_attr(
     ):
+        # TODO: Constant tensors and buffer/weights are both categorized into `get_attr`,
+        # but they are different to ONNX. We need to distinguish them.
+        # Constant tensors are inputs of the graph, and buffer/weights are initializers.
```
```suggestion
        # Constant tensors should become ONNX constants in the graph, while buffers/weights ONNX initializers.
        # For now they are all converted to ONNX initializers.
```
Is this more accurate?

BowenBao commented on torch/onnx/_internal/exporter.py:
```diff
@@ -954,10 +954,22 @@ def export(self) -> ExportOutput:
                 op_level_debug=self.options.op_level_debug,
             )
 
+            # NOTE: Filter out the initializers with fake tensors when it's fake_mode exporting,
+            # otherwise, the ONNX exporter will fail: RuntimeError: basic_string::_M_construct null
+            # not valid
```
```suggestion
            # NOTE: Filter out the initializers with fake tensors when it's fake_mode exporting.
            # Otherwise, the ONNX exporter will fail: RuntimeError: basic_string::_M_construct null
            # not valid.
            # Concrete data is expected to be filled for those initializers later during `ExportOutput.save`.
```

BowenBao reviewed (approved):
LG w/ comment

titaiwangms committed:
Update on "[ONNX] Support constant tensors in FakeMode exporting"


Fixes https://github.com/pytorch/pytorch/issues/107475

- Constant tensors was wrongly recognized as weights and buffers, and then was detached from its default value during `to_model_proto`. This PR fixes the bug and pick up Bloom CI test back successfully. NOTE: non-persistent buffer and weights has different situation and is not fixed by this PR.
- Reduce transformers model size by modifying their config parameters to speed up CI tests. (Unrelated to this PR title)

Corresponding change with https://github.com/microsoft/onnxscript/pull/1023

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge
